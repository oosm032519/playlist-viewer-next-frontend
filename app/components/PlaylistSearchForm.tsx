// app/components/PlaylistSearchForm.tsx

"use client";

import {useState} from "react";
import {useForm} from "react-hook-form";
import {yupResolver} from "@hookform/resolvers/yup";
import {schema} from "../validationSchema";
import {useSearchPlaylists} from "../hooks/useSearchPlaylists";
import {Button} from "./ui/button";
import {Input} from "./ui/input";
import {
    Form,
    FormControl,
    FormField,
    FormItem,
    FormMessage,
} from "./ui/form";
import {Card, CardContent, CardHeader, CardTitle} from "./ui/card";
import LoadingSpinner from "./LoadingSpinner";
import {Playlist} from "../types/playlist";
import {useQueryClient} from "@tanstack/react-query";
import PaginationButtons from "./PaginationButtons";
import {usePlaylist} from "../context/PlaylistContext";

interface SearchFormInputs {
    query: string;
}

interface PlaylistSearchFormProps {
    onSearch(playlists: Playlist[]): void;
}

/**
 * プレイリスト検索フォームコンポーネント
 * @param {PlaylistSearchFormProps} props - コンポーネントのプロパティ
 * @returns {JSX.Element} プレイリスト検索フォーム
 */
export default function PlaylistSearchForm({
                                               onSearch,
                                           }: PlaylistSearchFormProps) {
    // 現在のページ番号を管理するステート
    const [currentPage, setCurrentPage] = useState(1);
    // 現在のプレイリストデータを管理するステート
    const [currentPlaylists, setCurrentPlaylists] = useState<Playlist[]>([]);
    // React Queryのクライアントを取得
    const queryClient = useQueryClient();
    
    const {selectedPlaylistId} = usePlaylist();
    
    // フォームの設定
    const form = useForm<SearchFormInputs>({
        resolver: yupResolver(schema),
        defaultValues: {query: ""},
    });
    
    // プレイリスト検索のミューテーションを設定
    const searchMutation = useSearchPlaylists((data) => {
        onSearch(data);
        setCurrentPlaylists(data);
    });
    
    /**
     * フォーム送信時の処理
     * @param {SearchFormInputs} data - フォームの入力データ
     */
    const onSubmit = async (data: SearchFormInputs) => {
        setCurrentPage(1);
        searchMutation.mutate({query: data.query, page: 1, limit: 20});
    };
    
    /**
     * 次のページボタン押下時の処理
     */
    const handleNextPage = () => {
        const nextPage = currentPage + 1;
        setCurrentPage(nextPage);
        
        // キャッシュからデータを取得
        const cachedData = queryClient.getQueryData([
            "playlists",
            form.getValues("query"),
            nextPage,
        ]) as Playlist[] | undefined;
        
        if (cachedData) {
            // キャッシュがある場合はキャッシュからデータを設定し、親コンポーネントに通知
            setCurrentPlaylists(cachedData);
            onSearch(cachedData);
        } else {
            // キャッシュがない場合はAPIリクエストを送信
            searchMutation.mutate({
                query: form.getValues("query"),
                page: nextPage,
                limit: 20,
            });
        }
    };
    
    
    /**
     * 前のページボタン押下時の処理
     */
    const handlePrevPage = () => {
        const prevPage = currentPage - 1;
        setCurrentPage(prevPage);
        
        // キャッシュからデータを取得
        const cachedData = queryClient.getQueryData([
            "playlists",
            form.getValues("query"),
            prevPage,
        ]) as Playlist[] | undefined;
        
        if (cachedData) {
            // キャッシュがある場合はキャッシュからデータを設定し、親コンポーネントに通知
            setCurrentPlaylists(cachedData);
            onSearch(cachedData);
        } else {
            // キャッシュがない場合はAPIリクエストを送信
            searchMutation.mutate({
                query: form.getValues("query"),
                page: prevPage,
                limit: 20,
            });
        }
    };
    
    return (
        <>
            <Card>
                <CardHeader>
                    <CardTitle>Enter Playlist Name</CardTitle>
                </CardHeader>
                <CardContent>
                    <Form {...form}>
                        <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
                            <FormField
                                control={form.control}
                                name="query"
                                render={({field}) => (
                                    <FormItem>
                                        <FormControl>
                                            <div className="flex space-x-2">
                                                <Input
                                                    placeholder="Enter playlist name"
                                                    {...field}
                                                    disabled={searchMutation.isPending}
                                                />
                                                <Button
                                                    type="submit"
                                                    disabled={searchMutation.isPending}
                                                >
                                                    {searchMutation.isPending ? "Searching..." : "Search"}
                                                </Button>
                                            </div>
                                        </FormControl>
                                        <FormMessage/>
                                    </FormItem>
                                )}
                            />
                        </form>
                    </Form>
                </CardContent>
            </Card>
            <LoadingSpinner loading={searchMutation.isPending}/>
            {currentPlaylists.length > 0 && !selectedPlaylistId &&( // 条件を追加
                <PaginationButtons
                    currentPage={currentPage}
                    isPending={searchMutation.isPending}
                    hasNextPage={currentPlaylists.length === 20}
                    onNextPage={handleNextPage}
                    onPrevPage={handlePrevPage}
                />
            )}
        </>
    );
}
