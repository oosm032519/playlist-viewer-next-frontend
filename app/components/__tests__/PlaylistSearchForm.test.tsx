// PlaylistSearchForm.test.tsx

import React from 'react';
import {render, screen, fireEvent, waitFor} from '@testing-library/react';
import axios from 'axios';
import PlaylistSearchForm from '../PlaylistSearchForm';
import '@testing-library/jest-dom';

// axiosのモック化
jest.mock('axios');

// テスト用のダミーデータ
const mockPlaylists = [
    {id: '1', name: 'Playlist 1'},
    {id: '2', name: 'Playlist 2'},
];

describe('PlaylistSearchForm', () => {
    // 各テストの前にモックをリセット
    beforeEach(() => {
        jest.resetAllMocks();
    });
    
    // 基本的なレンダリングテスト
    test('renders search input and button', () => {
        render(<PlaylistSearchForm onSearch={() => {
        }}/>);
        
        expect(screen.getByPlaceholderText('Enter playlist name')).toBeInTheDocument();
        expect(screen.getByRole('button', {name: 'Search'})).toBeInTheDocument();
    });
    
    // フォーム送信のテスト
    test('submits form with valid input', async () => {
        const onSearchMock = jest.fn();
        (axios.get as jest.Mock).mockResolvedValueOnce({data: mockPlaylists});
        
        render(<PlaylistSearchForm onSearch={onSearchMock}/>);
        
        const input = screen.getByPlaceholderText('Enter playlist name');
        const button = screen.getByRole('button', {name: 'Search'});
        
        fireEvent.change(input, {target: {value: 'test query'}});
        fireEvent.click(button);
        
        await waitFor(() => {
            expect(axios.get).toHaveBeenCalledWith('/api/playlists/search?query=test query');
            expect(onSearchMock).toHaveBeenCalledWith(mockPlaylists);
        });
    });
    
    // バリデーションエラーのテスト
    test('displays validation error for short input', async () => {
        render(<PlaylistSearchForm onSearch={() => {
        }}/>);
        
        const input = screen.getByPlaceholderText('Enter playlist name');
        const button = screen.getByRole('button', {name: 'Search'});
        
        fireEvent.change(input, {target: {value: 'a'}});
        fireEvent.click(button);
        
        await waitFor(() => {
            expect(screen.getByText('最低2文字以上入力してください')).toBeInTheDocument();
        });
    });
    
    // ローディング状態のテスト
    test('displays loading state during search', async () => {
        (axios.get as jest.Mock).mockImplementation(() => new Promise(() => {
        })); // 永続的な保留状態を模倣
        
        render(<PlaylistSearchForm onSearch={() => {
        }}/>);
        
        const input = screen.getByPlaceholderText('Enter playlist name');
        const button = screen.getByRole('button', {name: 'Search'});
        
        fireEvent.change(input, {target: {value: 'test query'}});
        fireEvent.click(button);
        
        await waitFor(() => {
            expect(button).toHaveTextContent('Searching...');
            expect(button).toBeDisabled();
            expect(input).toBeDisabled();
        });
    });
});
