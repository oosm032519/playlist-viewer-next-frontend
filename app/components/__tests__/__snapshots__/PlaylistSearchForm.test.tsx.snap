// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistSearchForm matches snapshot 1`] = `
[
  <div
    className="rounded-xl border bg-card text-card-foreground shadow"
  >
    <div
      className="p-6 pt-0"
    >
      <form
        className="space-y-8"
        onSubmit={[Function]}
      >
        <div
          className="space-y-2"
        >
          <div
            aria-describedby=":r0:-form-item-description"
            aria-invalid={false}
            className="flex space-x-2"
            id=":r0:-form-item"
          >
            <input
              className="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
              disabled={false}
              name="query"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Enter playlist name"
              value=""
            />
            <button
              className="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
              disabled={false}
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
  <div
    aria-label="Loading progress"
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={0}
    className="fixed top-0 left-0 w-screen h-screen flex items-center justify-center bg-black/50 z-50"
    role="progressbar"
    style={
      {
        "display": "none",
      }
    }
  />,
]
`;
